src/impy/kinematics.py

```diff
@@ -185,57 +100,60 @@ def __init__(
         if particle1 is None or particle2 is None:
             raise ValueError("particle1 and particle2 must be set")
 
-        self.p1 = _normalize_particle(particle1)
-        self.p2 = _normalize_particle(particle2)
+        self.p1 = process_particle(particle1)
+        self.p2 = process_particle(particle2)
 
         if isinstance(self.p1, CompositeTarget):
             raise ValueError("Only 2nd particle can be CompositeTarget")
 
         p2_is_composite = isinstance(self.p2, CompositeTarget)
 
-        m1 = mass(self.p1)
+        m1 = nucleon_mass if self.p1.is_nucleus else mass(self.p1)
         m2 = nucleon_mass if p2_is_composite or self.p2.is_nucleus else mass(self.p2)
 
         self.beams = (np.zeros(4), np.zeros(4))
 
-        # Input specification in center of mass frame
+        # Input specification in center-of-mass frame
         if ecm is not None:
-            self.frame = EventFrame.CENTER_OF_MASS if frame is None else frame
+            self.frame = frame or EventFrame.CENTER_OF_MASS
             self.ecm = ecm
-            self.elab = 0.5 * (ecm**2 - m1**2 - m2**2) / m2
+            self.elab = ecm2elab(ecm, m1, m2)
             self.plab = energy2momentum(self.elab, m1)
         # Input specification as 4-vectors
         elif beam is not None:
             if p2_is_composite:
                 raise ValueError("beam cannot be used with CompositeTarget")
-            self.frame = EventFrame.GENERIC if frame is None else frame
+            self.frame = frame or EventFrame.GENERIC
             p1, p2 = beam
             self.beams[0][2] = p1
             self.beams[1][2] = p2
-            self.beams[0][3] = np.sqrt(m1**2 + p1**2)
-            self.beams[1][3] = np.sqrt(m2**2 + p2**2)
+            self.beams[0][3] = momentum2energy(p1, m1)
+            self.beams[1][3] = momentum2energy(p2, m2)
             s = np.sum(self.beams, axis=0)
-            self.ecm = np.sqrt(s[3] ** 2 - np.sum(s[:3] ** 2))
-            self.elab = 0.5 * (self.ecm**2 - m1**2 + m2**2) / m2
+            # We compute ecm with energy2momentum. It is not really energy to momentum,
+            # but energy2momentum(x, y) computes x^2 - y^2, which is what we need. Here,
+            # I use that px and py are always zero, if we ever change this, many formulas
+            # have to change in this class, like all the boosts
```

Comment:
Not sure I understand this. The formulae seem identical to what was there before, except that before ecm was computed using the sum over px/y/z whereas now the code really assumes px and py to be zero. Or do you mean something else?

We don’t have the boosts in place to allow for arbitrary incidence and @jncots was working on a generic lorentz trafo, which at that time was downvoted because there is a different maintained lib. I still believe that we should allow for arbitrary kinematic specifications since most of the pieces are already in place for that. Also, we would need to do extensive testing (not unit testing) of an external lorentz trafo code, in particular to mitigate the unavoidable round-off errors for UHECR energies. We have some recipes from FLUKA how one can iteratively mitigate the round-off errors, preserving stuff like particles on-mass shell etc. Not that straight forward, though.

Reply:
This is just to use named formulas everywhere instead of raw computations, and `energy2momentum` uses a neat numerical trick that we can also use here. That's why I changed it. I can restore the np.sum over the momenta, but right now that's just wasting time, because px and py are always zero. We currently pass the beams via the z-components.

I think we should first fix the stuff that is already in the library and not working correctly, before adding any more features. The addition of CompositeTarget already caused a lot of additional issues, time that would have been better spend on making everything working that the library advertises to do.

Reply:
No no, it's fine as it is, no need to sum over px, py. I just wanted to know if there were "typos" that could have resulted in wrong formulas, which would impact already derived works.